onnx/shape_inference/implementation.cc

```diff
@@ -439,69 +459,67 @@ class ShapeInferenceImplBase {
         if (schema->has_type_and_shape_inference_function()) {
           schema->GetTypeAndShapeInferenceFunction()(ctx);
         } else if (schema->HasFunction()) {
-          processCall(n, *(schema->GetFunction()), ctx);
-        } else {
-          // Continue with inference for remaining nodes
-          return;
+          ProcessCall(n, *(schema->GetFunction()), ctx);
+        } // else: rely on schema->CheckInputOutputType() down below.
+        // check type-constraints specified via type variables
+        if (options.check_type) {
+          schema->CheckInputOutputType(ctx);
         }
       } else if (model_local_functions_map.size() > 0) {
         auto iter = model_local_functions_map.find(GetModelLocalFunctionsMapIdentifier(n.domain(), n.op_type()));
         if (iter != model_local_functions_map.end()) {
-          processCall(n, *(iter->second), ctx);
+          ProcessCall(n, *(iter->second), ctx);
         } else {
           has_unsupported_op = true;
-          return;
         }
       } else {
         has_unsupported_op = true;
-        return;
+      }
+      if (!has_unsupported_op) {
+        for (int i = 0; i < n.output_size(); ++i) {
+          // skip type and shape propagation for missing optional outputs.
+          if (!n.output(i).empty())
+            UpdateType(n.output(i), ctx.getOutputType(i));
+        }
+        // Constant values are tracked to improve inference/checking for subsequent nodes.
+        ProcessConstant(n);
+        // If data-propagation is enabled, partial-evaluation (aka data-propagation) is performed
+        // to improve inference/checking for subsequent nodes.
+        if (options.enable_data_propagation && schema && schema->has_data_propagation_function()) {
+          if (generated_shape_data_by_name == nullptr) {
+            fail_shape_inference(
+                "Container for generated shape data cannot be nullptr when enable_data_propagation option is set.");
```

Comment:
Was there an action we can recommend to users?

Reply:
This is pre-existing code, just reformatted here due to indentation. Will leave it as-is. In fact, I suspect this can be an assertion, because this check happens at constructor-time earlier.